### PR TITLE
Analytic example and tests for Greedy Kernel Points

### PR DESCRIPTION
**Pull request by tp832944**
### PR Type

- Code style update (formatting, local variables)
- Refactoring (no functional changes)
- Documentation content changes
- Tests

### Description

Import refactored code for Greedy Kernel Points from private fork by @db091756. I have already reviewed this, although I think I have made some documentation updates and array shape annotations.

Add an analytic example for Greedy Kernel Points and set up as an integration test.

Add an analytic unit test for the loss function for greedy_kernel_points_loss(). I have adapted this from work by @db091756.

Import test class `TestHelperFunctions` for coresubset solvers from private fork by @db091756. I have already reviewed this and have not made any changes.

### How Has This Been Tested?

Existing tests pass as expected.

Add an analytic example for Greedy Kernel Points and set up as an integration test.

Add an analytic unit test for the loss function for greedy_kernel_points_loss().

Import test class `TestHelperFunctions` for coresubset solvers.

### Does this PR introduce a breaking change?

Not to public API.

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have ensured my code is easy to understand, including docstrings and comments where necessary.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I have updated CHANGELOG.md, if appropriate.

### CODE REVIEW

**Comment by tp832944** [comment]
Failure in `TestMapReduce` on macOS appears to be unrelated.

**Review by gw265981 — COMMENTED** [review]
Happy with the changes overall, just a few minor comments. I have verified the analytic tests.

File: tests/unit/test_solvers.py [hunk 1]
```diff
@@ -2006,16 +2242,249 @@ def test_approximate_inverse(self, reduce_problem: _ReduceProblem) -> None:
...
+        candidate_coresets: jax.Array,
+        feature_gramian: jax.Array,
+        responses: jax.Array,
+        identity: jax.Array,
+        expect: jax.Array,
+    ) -> None:
+        r"""
+        Test _greedy_kernel_points_loss() analytically with zero regularisation.
+
+        In the first test case, we consider a dataset of size three, and we test the
```
[the remark this answers is not in the document]
Comment (tp832944) [reply tests/unit/test_solvers.py:2359]: Nope, but have added a comment explaining why not.

File: coreax/solvers/coresubset.py [hunk 2]
```diff
@@ -670,143 +681,513 @@ class GreedyKernelPointsState(eqx.Module):
...
 
     :param feature_gramian: Cached feature kernel gramian matrix, should be padded with
-        an additional row and column of zeros.
+        an additional row and column of zeroes.
     """
 
     feature_gramian: Array
 
 
+# Overload for the case where we want to construct both the identity array and the
```
[the remark this answers is not in the document]
Comment (tp832944) [reply coreax/solvers/coresubset.py:690]: I did mention having helper functions outside classes at the time to @db091756 but we decided to keep it this way for consistency with other methods in this submodule.

[the remark this answers is not in the document]
Comment (tp832944) [reply coreax/solvers/coresubset.py:690]: https://github.com/gchq/coreax/issues/1008

File: tests/unit/test_solvers.py [hunk 3]
```diff
@@ -1966,7 +2202,7 @@ def refine_problem(
             expected_coresubset = None
         else:
             raise ValueError("Invalid fixture parametrization")
-        initial_coresubset = Coresubset.build(indices, dataset)
+        initial_coresubset = Coresubset(indices, dataset)
```
[the remark this answers is not in the document]
Comment (tp832944) [reply tests/unit/test_solvers.py]: I did make this change but have now reverted it. `indices` is already an instance of `Data`, so this is correct. `build` is for use when `indices` is an `Array`.

[the remark this answers is not in the document]
Comment (gw265981) [reply tests/unit/test_solvers.py]: This makes sense. Using .build should give the same result but I'm happy with this approach.

**Review by gw265981 — APPROVED** [review]
